### PR TITLE
Add support for chained schemas

### PR DESCRIPTION
**Pull request by elprans**
The new `ChainedSchema` implementation added here takes two regular
schemas and chains them together a-la `ChainMap`.  This is necessary to
separate the read-only `std` portion of schema from user-defined schema
in order to save on serialization cost of transmitting the schema over
RPC.

### CODE REVIEW

File: edb/edgeql/compiler/setgen.py [hunk 1]
```diff
@@ -783,7 +783,7 @@ def type_intersection_set(
     arg_type = get_set_type(source_set, ctx=ctx)
 
     result = schemactx.apply_intersection(arg_type, stype, ctx=ctx)
-    if result.stype is arg_type:
+    if result.stype == arg_type:
```
Comment (1st1) [inline edb/edgeql/compiler/setgen.py:786]: Ouch. This is a tricky thing to debug. Is it possible to tell mypy to prohibit using `is` on certain types?

File: edb/schema/objects.py [hunk 2]
```diff
@@ -817,8 +820,8 @@ def get_verbosename(
         dname = self.get_displayname(schema)
         return f"{clsname} '{dname}'"
 
-    def __init__(self, *, _private_init: bool) -> None:
-        pass
+    def __init__(self, *, _private_init: uuid.UUID) -> None:
```
Comment (1st1) [inline edb/schema/objects.py:823]: I'd call it "private_id"

```suggestion
    def __init__(self, *, _private_id: uuid.UUID) -> None:
```

File: edb/schema/schema.py [hunk 3]
```diff
@@ -1004,19 +1284,401 @@ def __init__(
...
+        self,
+        obj_id: uuid.UUID,
+        default: Union[so.Object_T, so.NoDefaultT, None] = so.NoDefault,
+        *,
+        type: Optional[Type[so.Object_T]] = None,
+    ) -> Optional[so.Object_T]:
+        try:
+            return self._top_schema.get_by_id(
+                obj_id, type=type)
+        except errors.InvalidReferenceError:
```
Comment (1st1) [inline edb/schema/schema.py:1535]: Maybe we optimize this one too by checking if the `_top_schema` has this id? Exceptions are expensive.

File: edb/schema/schema.py [hunk 4]
```diff
@@ -1004,19 +1284,401 @@ def __init__(
...
+
+    def get_global(  # NoQA: F811
+        self,
+        objtype: Type[so.Object_T],
+        name: str,
+        default: Union[so.Object_T, so.NoDefaultT, None] = so.NoDefault,
+    ) -> Optional[so.Object_T]:
+        try:
+            return self._top_schema.get_global(objtype, name)
+        except errors.InvalidReferenceError:
```
Comment (1st1) [inline edb/schema/schema.py:1566]: Ditto, optimize this one too? Checking `_globalname_to_id` on `top_schema` should do it

**Review by 1st1 — APPROVED** [review]

File: edb/edgeql/compiler/setgen.py (hunk 1 above)
Comment (elprans) [reply edb/edgeql/compiler/setgen.py:786]: mypy has to grow a few extra hooks and then we'll be able to do this with a plugin.

Comment (1st1) [reply edb/edgeql/compiler/setgen.py:786]: Summoning @msullivan

**Pull request merged**